src/mpid/ch4/shm/xpmem/xpmem_init.c

```diff
@@ -51,21 +51,46 @@ int MPIDI_XPMEM_mpi_init_hook(int rank, int size, int *n_vcis_provided, int *tag
         MPIDI_XPMEM_global.segmaps[i].apid = -1;        /* get apid at the first communication  */
 
         /* Init AVL tree based segment cache */
-        MPIDI_XPMEM_segtree_init(&MPIDI_XPMEM_global.segmaps[i].segcache);
+        MPIDI_XPMEM_segtree_init(&MPIDI_XPMEM_global.segmaps[i].segcache_user);
+        MPIDI_XPMEM_segtree_init(&MPIDI_XPMEM_global.segmaps[i].segcache_cnt);
     }
 
     /* Check to make sure all processes initialized XPMEM correctly. */
     if (anyfail) {
         for (i = 0; i < num_local; i++) {
-            MPIDI_XPMEM_segtree_delete_all(&MPIDI_XPMEM_global.segmaps[i].segcache);
+            MPIDI_XPMEM_segtree_delete_all(&MPIDI_XPMEM_global.segmaps[i].segcache_user);
+            MPIDI_XPMEM_segtree_delete_all(&MPIDI_XPMEM_global.segmaps[i].segcache_cnt);
         }
         /* Not setting an mpi_errno value here because we can handle the failure of XPMEM
          * gracefully. */
         goto fn_fail;
     }
 
     /* Initialize other global parameters */
+    MPIR_CHKPMEM_MALLOC(MPIDI_XPMEM_global.coop_counter_direct, MPIDI_XPMEM_cnt_t **,
+                        sizeof(MPIDI_XPMEM_cnt_t *) * num_local,
+                        mpi_errno, "xpmem direct counter array", MPL_MEM_SHM);
+    MPIR_CHKPMEM_MALLOC(MPIDI_XPMEM_global.coop_counter_origin_direct, MPIDI_XPMEM_seg_t **,
+                        sizeof(MPIDI_XPMEM_seg_t *) * num_local,
+                        mpi_errno, "xpmem direct counter segs", MPL_MEM_SHM);
+
     MPIDI_XPMEM_global.sys_page_sz = (size_t) sysconf(_SC_PAGESIZE);
+    uint64_t cnt_mem_addr = (uint64_t) (&MPIDI_XPMEM_cnt_mem_direct);
```

Comment:
Can you reorder the initialization code as below? The current comment does not match the code.
```
/* Initialize other global parameters */
MPIDI_XPMEM_global.sys_page_sz = (size_t) sysconf(_SC_PAGESIZE);

/* Initialize coop counter */
...
```